Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximated_centroidal_Voronoi_diagram_remeshing.h

```diff
@@ -0,0 +1,1319 @@
+
+        vertex_queue.push(vd);
+        put(vertex_visited_pmap, vd, true);
+        while (!vertex_queue.empty())
+        {
+          Vertex_descriptor v = vertex_queue.front();
+          vertex_queue.pop();
+          cluster_components[c].back().push_back(v);
+
+          for (Halfedge_descriptor hd : halfedges_around_source(v, pmesh))
+          {
+            Vertex_descriptor v2 = target(hd, pmesh);
+            int c2 = get(vertex_cluster_pmap, v2);
+            if (c2 == c && !get(vertex_visited_pmap, v2))
+            {
+              vertex_queue.push(v2);
+              put(vertex_visited_pmap, v2, true);
+            }
+          }
+        }
+      }
+    }
+
+    // loop over clusters to delete disconnected components except the largest one
+    for (int c = 0; c < nb_clusters; ++c)
+    {
+      if (cluster_components[c].size() <= 1) continue; // only one component, no need to do anything
+      ++nb_disconnected;
+      std::size_t max_component_size = 0;
+      std::size_t max_component_index = -1;
+      for (std::size_t component_i = 0; component_i < cluster_components[c].size(); ++component_i)
+      {
+        if (cluster_components[c][component_i].size() > max_component_size)
+        {
+          max_component_size = cluster_components[c][component_i].size();
+          max_component_index = component_i;
+        }
+      }
+      // set cluster to -1 for all components except the largest one
+      for (std::size_t component_i = 0; component_i < cluster_components[c].size(); ++component_i)
+      {
+        if (component_i != max_component_index)
+        {
+          for (Vertex_descriptor vd : cluster_components[c][component_i])
+          {
+            put(vertex_cluster_pmap, vd, -1);
+            // remove vd from cluster c
+            typename GT::Point_3 vp = get(vpm, vd);
+            typename GT::Vector_3 vpv(vp.x(), vp.y(), vp.z());
+            clusters[c].remove_vertex(vpv, get(vertex_weight_pmap, vd), get(vertex_quadric_pmap, vd));
+            // add all halfedges around v except hi to the queue
+            for (Halfedge_descriptor hd : halfedges_around_source(vd, pmesh))
+            {
+              // add hd to the queue if its target is not in the same cluster
+              Vertex_descriptor v2 = target(hd, pmesh);
+              int c2 = get(vertex_cluster_pmap, v2);
+              if (c2 != c)
+                clusters_edges_active.push(hd);
+            }
+          }
+        }
+      }
+    }
+
+    // dump_mesh_with_cluster_colors(pmesh, vertex_cluster_pmap, "/tmp/clusters_loop_"+std::to_string(nb_loops)+".ply");
+
+    std::cout << "# nb_disconnected: " << nb_disconnected << "\n";
+    ++nb_loops;
+
+  } while (nb_disconnected > 0 || nb_loops < 3 );
+
+  /// Construct new Mesh
+  std::vector<std::size_t> valid_cluster_map(nb_clusters, -1);
+  std::vector<typename GT::Point_3> points;
+
+  std::vector<std::array<std::size_t, 3>> polygons;
+  TriangleMesh simplified_mesh;
```

Comment:
unused, but if you guarantee that the soup is a mesh, what's the point of returning a soup from _impl anyway?

Reply:
_impl no longer makes sense anyway. I'll remove it